### PR TITLE
 [LI-HOTFIX] Log the partition being made online due to unclean leader election.

### PR DESCRIPTION
**Pull request by smccauliff**
Add additional unit tests for unclean leader election.

### CODE REVIEW

File: core/src/main/scala/kafka/controller/Election.scala [hunk 1]
```diff
@@ -17,11 +17,12 @@
 package kafka.controller
 
 import kafka.api.LeaderAndIsr
+import kafka.utils.Logging
 import org.apache.kafka.common.TopicPartition
 
 case class ElectionResult(topicPartition: TopicPartition, leaderAndIsr: Option[LeaderAndIsr], liveReplicas: Seq[Int])
 
-object Election {
+object Election extends Logging {
```
Comment (radai-rosenblatt) [inline core/src/main/scala/kafka/controller/Election.scala:25]: how often will this object be instantiated? reason im asking is we've been burned by the Logging aspect on short lived objects before

File: core/src/main/scala/kafka/controller/PartitionStateMachine.scala [hunk 2]
```diff
@@ -431,19 +432,26 @@ class ZkPartitionStateMachine(config: KafkaConfig,
...
+                        case Some(uncleanReplicaId) => Some(uncleanReplicaId, true)
+                        case None => None
+                      }
+                  } else {
+                    None
+                  }
     }
   }
 
+
```
Comment (radai-rosenblatt) [inline core/src/main/scala/kafka/controller/PartitionStateMachine.scala]: nit - whitespace

File: core/src/main/scala/kafka/controller/PartitionStateMachine.scala [hunk 3]
```diff
@@ -355,6 +355,7 @@ class ZkPartitionStateMachine(config: KafkaConfig,
     val (invalidPartitionsForElection, validPartitionsForElection) = leaderIsrAndControllerEpochPerPartition.partition { case (_, leaderIsrAndControllerEpoch) =>
       leaderIsrAndControllerEpoch.controllerEpoch > controllerContext.epoch
     }
+    //TODO: if we find any invalid partitions for election shouldn't the controller thread exit since there is a new controller?
```
Comment (xiowu0) [inline core/src/main/scala/kafka/controller/PartitionStateMachine.scala]: not sure if we want to do it inside PartitionStateMachine.scala, maybe the controller change event has already been en-queued.

**Review by xiowu0 — APPROVED** [review]
LGTM. Thanks for the patch

File: core/src/main/scala/kafka/controller/Election.scala (hunk 1 above)
Comment (smccauliff) [reply core/src/main/scala/kafka/controller/Election.scala:25]: It's an object so it should  only be instantiated once.

File: core/src/main/scala/kafka/controller/PartitionStateMachine.scala (hunk 3 above)
Comment (smccauliff) [reply core/src/main/scala/kafka/controller/PartitionStateMachine.scala]: Removed TODO

**Pull request merged**